src/op_analytics/cli/subcommands/pulls/agora/delegates.py

```diff
@@ -0,0 +1,213 @@
+                    self.max_offset = min(upper_bound)
+                elif lower_bound:
+                    self.max_offset = max(lower_bound)
+                else:
+                    break
+                return self.max_offset
+
+            resp_lo = self.request(lo)
+            resp_hi = self.request(hi)
+
+            if resp_lo is None or resp_hi is None:
+                log.warning(f"Failed to fetch data for offsets {lo} or {hi}.")
+                break
+
+            print(f"max offset search range {lo} -- {hi}")
+
+            diff = hi - lo
+            midpoint = lo + (diff) // 2
+
+            if resp_lo.has_next and resp_hi.has_next:
+                lower_bound.add(hi)
+                lo = max(lower_bound)
+                hi = min(upper_bound) if upper_bound else hi * 10
+            elif resp_lo.has_next and not resp_hi.has_next:
+                lower_bound.add(lo)
+                upper_bound.add(hi)
+                if diff < 1000:
+                    self.min_offset = lo
+                    self.max_offset = hi
+                    return hi
+                else:
+                    hi = midpoint
+            else:
+                log.warning("Unexpected state encountered in binary search.")
+                hi = midpoint
+
+    def concurrent_fetch(self):
+        self.max_offset = self.binary_search_max_offset()
+        offsets = list(range(0, self.max_offset, self.limit))[:10]
+        data = run_concurrently(
+            function=lambda offset: self.request(offset).data,
+            targets=offsets,
+            max_workers=self.max_workers,
+        )
+        return list(itertools.chain.from_iterable(data.values()))
+
+
+def pull_delegates():
+    """Pull and write agora delegates data."""
+    p = Paginator(url=f"{BASE_URL}/delegates", params={"sort": "voting_power"}, max_workers=32)
+    delegate_data = p.concurrent_fetch()
+
+    # Clean and transform to Pandas
+    df = pd.json_normalize(delegate_data, sep="_")
+    df.columns = df.columns.map(camelcase_to_snakecase)
+    df = df.convert_dtypes()
+    # In order to be able to drop duplicates we cannot have iterables in the dataframe
+    df = df.map(lambda x: str(x) if isinstance(x, list) else x)
+    # Check for duplicates
+    if df.shape[0] != len(df["address"].drop_duplicates()):
+        df = df.drop_duplicates()
+        log.warning(
+            f"Found duplicates in delegates data. Dropped {df.shape[0] - len(df['address'].drop_duplicates())} rows."
+        )
+    # Convert voting power columns to numeric
+    for col in ["voting_power_total", "voting_power_direct", "voting_power_advanced"]:
+        df[col] = pd.to_numeric(df[col], errors="coerce")
+
+    df["dt"] = (
+        now_dt()
+    )  # Todo: This is to partition by pull date. Figure out if this is the correct way to do this.
+    # Convert to polars
+    df = pl.from_pandas(df)
+
+    # Split to delegates that have voting power and those that don't
```

Comment:
not sure about why splitting into two tables. Could we get away with a single table and having readers do the filtering?

Reply:
Fixed as mentioned above